hathor/transaction/storage/transaction_storage.py

```diff
@@ -988,6 +998,58 @@ def _topological_fast(self) -> Iterator[BaseTransaction]:
         yield from cur_blocks
         yield from cur_txs
 
+    def _topological_fast2(self) -> Iterator[BaseTransaction]:
+        import heapq
+        from dataclasses import dataclass, field
+        cur_timestamp = settings.GENESIS_TIMESTAMP
+
+        @dataclass(order=True)
+        class Item:
+            timestamp: int
+            is_transaction: bool
+            tx: BaseTransaction = field(compare=False)
+
+            def __init__(self, tx: BaseTransaction):
+                self.timestamp = tx.timestamp
+                self.is_transaction = tx.is_transaction
+                self.tx = tx
+
+        to_visit: List[Item] = list(map(Item, self.get_all_genesis()))
+        heapq.heapify(to_visit)
+        visited: Set[bytes] = set()
+        to_check: List[Item] = []
+        to_yield: List[Item] = []
+        while (to_visit or to_check or to_yield):
+            while (to_visit and to_visit[0].timestamp <= cur_timestamp) or (not to_check):
+                pop_item = heapq.heappop(to_visit)
+                heapq.heappush(to_check, pop_item)
+                new_children = set(pop_item.tx.get_metadata().children) - visited
+                visited.update(new_children)
+                for child in new_children:
+                    try:
+                        child_tx = self.get_transaction(child)
+                    except TransactionDoesNotExist:
```

Comment:
If we have the metadata, shouldn't we have the transaction?

Reply:
Ok. This case you described might be a common cause of corruption. Maybe we can fix it during initialization in a separate PR. We could use a similar structure to journaling. Before adding the tx to the db and the indexes, we could add it to a journaling db. After the tx is successfully added, we remove it from the the journal. When the full node starts, we check the journal and remove all partially added tx. What do you think?

Reply:
I think that would be great, and fairly independent from this. I'll link to this suggestion from the tracking issue that we have.

Reply:
Here it is: https://github.com/HathorNetwork/hathor-core/issues/411#issuecomment-1128825322